src/main/java/teammates/ui/webapi/action/GetStudentAction.java

```diff
@@ -17,37 +23,66 @@ protected AuthType getMinAuthLevel() {
+            String studentEmail = getNonNullRequestParamValue(Const.ParamsNames.STUDENT_EMAIL);
+            student = logic.getStudentForEmail(courseId, studentEmail);
+            if (student == null) {
+                throw new UnauthorizedAccessException(Const.ErrorMessages.UNAUTHORIZED_ACCESS);
+            }
+
+            InstructorAttributes instructor = logic.getInstructorForGoogleId(courseId, userInfo.id);
+            gateKeeper.verifyAccessible(instructor, logic.getCourse(courseId), student.section,
+                    Const.ParamsNames.INSTRUCTOR_PERMISSION_VIEW_STUDENT_IN_SECTIONS);
+        } else if (userInfo.isStudent) {
+            student = logic.getStudentForGoogleId(courseId, userInfo.id);
+            if (student == null) {
+                throw new UnauthorizedAccessException(Const.ErrorMessages.UNAUTHORIZED_ACCESS);
+            }
+
+            gateKeeper.verifyAccessible(student, course);
+        } else {
+            Optional<StudentAttributes> studentCheck = getUnregisteredStudent();
+            if (studentCheck.isPresent()) {
+                student = studentCheck.get();
+                gateKeeper.verifyAccessible(student, course);
+            } else {
+                throw new UnauthorizedAccessException(Const.ErrorMessages.UNAUTHORIZED_ACCESS);
+            }
         }
     }
 
     @Override
     public ActionResult execute() {
-        Intent intent = Intent.valueOf(getNonNullRequestParamValue(Const.ParamsNames.INTENT));
-        switch (intent) {
-        case STUDENT_SUBMISSION:
-            String courseId = getNonNullRequestParamValue(Const.ParamsNames.COURSE_ID);
-            StudentAttributes studentAttributes = getStudentOfCourseFromRequest(courseId);
-            return new JsonResult(new StudentInfo.StudentResponse(studentAttributes));
-        case FULL_DETAIL:
-            // TODO implement this when necessary
-            return null;
-        default:
-            throw new InvalidHttpParameterException("Unknown intent " + intent);
+        String courseId = getNonNullRequestParamValue(Const.ParamsNames.COURSE_ID);
+        StudentAttributes student = null;
+
+        if (userInfo == null) {
+            Optional<StudentAttributes> studentCheck = getUnregisteredStudent();
+            if (studentCheck.isPresent()) {
+                student = studentCheck.get();
+            }
+        } else if (userInfo.isInstructor) {
+            String studentEmail = getNonNullRequestParamValue(Const.ParamsNames.STUDENT_EMAIL);
+            student = logic.getStudentForEmail(courseId, studentEmail);
+        } else if (userInfo.isStudent) {
+            student = logic.getStudentForGoogleId(courseId, userInfo.id);
         }
-    }
 
+        if (student == null) {
+            return new JsonResult("No student found", HttpStatus.SC_NOT_FOUND);
+        }
+
+        StudentData studentData = new StudentData(student);
+
+        if (userInfo == null || userInfo.isStudent) {
```

Comment:
I think you should hide the data in the branches rather than here.

A user can be both a student and an instructor of the same course.

Reply:
Its tricky to hide in the branches (have to pull StudentData back and duplicate code). I changed the boolean condition to `!userInfo.isInstructor`. Let me know if there is a better way to do this.